onnxruntime/core/providers/qnn/qnn_execution_provider.cc

```diff
@@ -838,10 +935,22 @@ Status QNNExecutionProvider::Compile(const std::vector<FusedNodeAndGraph>& fused
       // fused node name is QNNExecutionProvider_QNN_[hash_id]_[id]
       // the name here must be same with context->node_name in compute_info
       qnn_models_.emplace(graph_meta_id, std::move(qnn_model));
+      qnn_models.erase(key);
 
       ORT_RETURN_IF_ERROR(CreateComputeFunc(node_compute_funcs, logger));
     }
 
+    if (share_ep_contexts_ && qnn_models.size() > 0) {
+      std::vector<std::shared_ptr<qnn::QnnModel>> shared_qnn_models;
+      for (auto& [key, value] : qnn_models) {
+        shared_qnn_models.push_back(std::move(qnn_models[key]));
+      }
+      std::string duplicate_graph_names;
+      bool has_duplicate_graph = SharedContext::GetInstance().SetSharedQnnModel(shared_qnn_models,
+                                                                                duplicate_graph_names);
+      ORT_RETURN_IF(has_duplicate_graph, "Duplicate graph names detect across sessions: " + duplicate_graph_names);
```

Comment:
Would this cause an issue for a context binary that has more than 2 graphs? 

For example, if we assume a single context binary has 3 graphs:
- Session 1 loads the 3 qnn graphs from the context binary. It only needs GRAPH_1, so it adds GRAPH_2 and GRAPH_3 to the shared state.
- Session 2 loads the 3 qnn graphs from the context binary. It only needs GRAPH_2, so it tries to add GRAPH_1 and GRAPH_3 to the shared state. This would result in an error because GRAPH_3 has already been added before.
- We wouldn't get this far, but now Session 3 would also load the 3 qnn graphs from the context binary and try to add duplicates again.

Reply:
From your description, it seems that session1 wants graph1, session2 wants graph2, session3 want graph3. OK. It works like this.
Session1 loads the context binary, and gets 3 graphs. It only wants graph1. So it put graph2 & graph3 to the shared state.
Session2 won't load the context binary again. It checks the shared state has graph2 which is required by the model it loads (the model has EPContext node with graph name for graph2). So, it directly get it from the shared state. That's it.
Same for session3.

Reply:
I see thanks.